server/src/main/java/io/crate/expression/reference/doc/lucene/ColumnFieldVisitor.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.expression.reference.doc.lucene;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.StoredFieldVisitor;
+import org.jetbrains.annotations.NotNull;
+
+import io.crate.common.collections.Maps;
+import io.crate.execution.dml.ArrayIndexer;
+import io.crate.metadata.ColumnIdent;
+import io.crate.metadata.Reference;
+import io.crate.metadata.doc.DocTableInfo;
+import io.crate.metadata.doc.SysColumns;
+import io.crate.types.ArrayType;
+import io.crate.types.DataType;
+
+/**
+ * Loads column values from stored fields and presents them as a java Map
+ */
+public class ColumnFieldVisitor extends StoredFieldVisitor {
+
+    private final Map<String, Field> fields = new HashMap<>();
+    private final Set<ColumnIdent> droppedColumns;
+    private final SourceParser storedSourceParser;
+
+    /**
+     * Creates a new ColumnFieldVisitor for the given table
+     */
+    public ColumnFieldVisitor(DocTableInfo table) {
+        this.droppedColumns
+            = table.droppedColumns().stream().map(Reference::column).collect(Collectors.toUnmodifiableSet());
+        this.storedSourceParser = new SourceParser(table);
+    }
+
+    private record Field(DataType<?> dataType, ColumnIdent column) implements Comparable<Field> {
+
+        public Object decode(SourceParser sourceParser, byte[] v) {
+            return dataType.storageSupportSafe().decodeFromBytes(column, sourceParser, v);
```

Comment:
Probably worth assigning the storageSupport instance to a property of `Field`. Seems like decode is called per doc - and StorageSupport instances can be created on access (e.g. for NumericStorage)